### PR TITLE
[PW-401] fix job cancelation

### PR DESCRIPTION
**Pull request by alexdicianu**
**Problem #1:** The existing implementation is trying to identify which jobs to cancel by including the entire list of labels in the filters. This is problematic because some labels are unique to each job (e.g. task id). Also, this makes it hard to implement fine grained restrictions, say at the site or environment level.

**Solution:** expose a new property `concurrency_policy_labels` which allows for a more fine-grained control of which labels are used to identify cancelable jobs. This will also replace `prevent_concurrent_labels` which can only used in combination with a specific concurrency policy.

**Problem #2:** When asking the job-runner to cancel a job, we need to use the backend (i.e. Cloud Build) job ID not the PostgreSQL ID.

**Solution:** switch to using the backend job ID. This restricts our cancellation ability to jobs which are currently running (statuses: QUEUED, WORKING). We are, thus, excluding the REQUESTED status by design which means we can have the job-manager publish a job request on the topic (status: REQUESTED) but not have the ability to cancel it until it is actually started. This could create some unwanted scenarios of customers push commits in this, arguably small, timeframe.

### CODE REVIEW

File: pkg/db/jobs.go [hunk 1]
```diff
@@ -137,23 +142,38 @@ func (q *internalQueryer) GetJobByID(ctx context.Context, jobID string) (*JobEnt
...
+			labelsArgs[k] = k + "." + v
+		}
+		labelsCondition = strings.Join(conditions, " AND ")
+	}
+	statusCondition := ""
+	if len(statuses) > 0 {
+		placeHolders := []string{}
+		for _, status := range statuses {
+			// TODO (!): parametrize this to prevent SQL injections. Fucking syntax errors!!
+			placeHolders = append(placeHolders, fmt.Sprintf("'%s'", status))
```
Comment (alexdicianu) [inline pkg/db/jobs.go:166]: Do not forget to look into this!

File: pkg/server/server.go [hunk 2]
```diff
@@ -202,16 +203,28 @@ func (s *Server) CreateJob(w http.ResponseWriter, r *http.Request, ps httprouter
...
 	}
 
 	ctx, cancel := ctxTimeout()
 	defer cancel()
 	ctx = trace.AddTraceFromHeaderToCtx(ctx, r.Header)
 
-	jobsList, err := s.App.ListJobs(ctx, singleValues)
+	limitStr, found := labels["limit"]
+	if !found {
+		limitStr = "0"
```
Comment (alexdicianu) [inline pkg/server/server.go:217]: This kinda bugs me. We should not be able to select all rows from the db without any limit.

File: pkg/app/app.go [hunk 3]
```diff
@@ -146,42 +157,58 @@ func (a *internalApp) GetJobLogs(ctx context.Context, jobID string, params map[s
...
+		jobs.StatusWorking,
+	}
+	jobsList, err := a.jobsQueryer.ListJobs(ctx, lm, statuses, 0)
 	if err != nil {
 		return err
 	}
 	log.Infof("planning to cancel %d jobs", len(jobsList))
 	eg, cctx := errgroup.WithContext(ctx)
+	// Limiting concurrency to at most 10 workers to avoid memory leaks or runaways.
+	eg.SetLimit(10)
```
Comment (IvanProdaiko) [inline pkg/app/app.go:183]: that's nice!

File: pkg/db/jobs.go (hunk 1 above)
Comment (alexdicianu) [reply pkg/db/jobs.go]: Yep, fixed it.

**Review by IvanProdaiko — APPROVED** [review]

**Pull request merged**